File: playground/ElixirAppHost/apphost.exs
# Aspire Elixir AppHost - Playground
# For more information, see: https://aspire.dev
#
# Elixir AppHost support is experimental. aspire.config.json turns the feature on
# for this directory. To turn it on for every AppHost, run:
#
#     aspire config set features:experimentalPolyglot:elixir true --global
#
# To run:
#
#     aspire run
#
# The applications live in ../ElixirApps. That directory also holds a C# AppHost
# that builds the same model.

Code.require_file(".aspire/modules/aspire.ex", __DIR__)

alias Aspire.DistributedApplicationBuilder, as: Builder
alias Aspire.Elixir.ElixirAppResource
alias Aspire.Elixir.PhoenixAppResource
alias Aspire.PostgreSQL.PostgresServerResource

builder = Aspire.create_builder!()

# The compute environment gives `aspire publish` a target. It writes a Docker Compose
# file and one Dockerfile for each Elixir application.
Builder.add_docker_compose_environment!(builder, "compose")

appdb =
  builder
  |> Builder.add_postgres!("db")
  |> PostgresServerResource.add_database!("appdb")

cache = Builder.add_redis!(builder, "cache")

# A fluent function returns the declared base handle, not the resource handle. The base
# handle modules hold no functions, so a plain pipe stops after one step. `tap/2` keeps
# the resource struct in the pipe. See "Known limits" in README.md.
builder
|> Builder.add_phoenix_app!("web", "../ElixirApps/phoenix_web")
|> tap(&PhoenixAppResource.with_ecto_database!(&1, appdb))
|> tap(&PhoenixAppResource.with_ecto_migrate!/1)
|> tap(&PhoenixAppResource.with_external_http_endpoints!/1)

builder
|> Builder.add_elixir_app!("worker", "../ElixirApps/worker")
|> tap(&ElixirAppResource.with_reference!(&1, cache))
|> tap(&ElixirAppResource.wait_for!(&1, cache))

builder
|> Aspire.build!()
|> Aspire.run!()

File: playground/DartAppHost/apphost.dart
// Aspire Dart AppHost - Playground
// For more information, see: https://aspire.dev
//
// Dart AppHost support is experimental. aspire.config.json turns the feature on
// for this directory. To turn it on for every AppHost, run:
//
//     aspire config set features:experimentalPolyglot:dart true --global
//
// To run:
//
//     aspire run
//
// The applications live in ../DartApps. That directory also holds a C# AppHost
// that builds the same model.

import '.aspire/modules/aspire.dart';

Future<void> main(List<String> args) async {
  final builder = await createBuilder(args);

  // The compute environment gives `aspire publish` a target. It writes a Docker
  // Compose file and one Dockerfile for each Dart application.
  await builder.addDockerComposeEnvironment('compose');

  final pg = await builder.addPostgres('pg');
  final appdb = await pg.addDatabase('appdb');
  final cache = await builder.addRedis('cache');

  final api = await builder.addServerpodApp(
      'api', '../DartApps/serverpod_api/serverpod_api_server');
  await api.withServerpodDatabase(appdb);
  await api.withServerpodRedis(cache);
  await api.withExternalHttpEndpoints();

  final site = await builder.addJasprApp('site', '../DartApps/jaspr_site');
  await site.withExternalHttpEndpoints();

  final worker = await builder.addDartApp('worker', '../DartApps/worker');
  await worker.withReference(api);
  await worker.waitFor(api);

  final app = await builder.build();
  await app.run();
}
